tests/toolbox_image.sh

```diff
@@ -13,6 +24,55 @@ result=0
 
 exit_on_failure=false
 
+format_dummy_partition()
+{
+    command="${1}"
+    sector_start="${2}"
+    device_file="${3}"
+
+    loop_device="$(chroot "${rootfs_dir}" /sbin/losetup -f)"
+    echo "Using next available loop device '${loop_device}' to format partition using: '${command}':"
+
+    touch "${rootfs_dir}${loop_device}"
+    mount --bind "${loop_device}" "${rootfs_dir}${loop_device}"
+
+    chroot "${rootfs_dir}" /sbin/losetup -o "${sector_start}" "${loop_device}" "${device_file}"
+    chroot "${rootfs_dir}" "${command}" "${loop_device}"
+    chroot "${rootfs_dir}" /sbin/losetup -d "${loop_device}"
+
+    if grep "${loop_device}" "/proc/mounts"; then
+        umount "${rootfs_dir}${loop_device}"
+        unlink "${rootfs_dir}${loop_device}"
+    fi
+}
+
+create_dummy_storage_device()
+{
+    partition_table_file="${rootfs_dir}/tmp/partition_table"
+
+cat <<EOT >> "${partition_table_file}"
+unit: sectors
+
+boot    : start=${BOOT_START},      size=$(( ROOTFS_START - BOOT_START )),    Id=83
+rootfs  : start=${ROOTFS_START},    size=$(( SPOOL_START - ROOTFS_START )),   Id=83
+spool   : start=${SPOOL_START},     size=$(( SECTOR_TOTAL - SPOOL_START )),   Id=83
+EOT
+
+    echo "Creating test image: '${rootfs_dir}${STORAGE_DEVICE_FILE}'"
+    dd if="/dev/zero" of="${rootfs_dir}${STORAGE_DEVICE_FILE}" bs="1" count="0" \
+        seek="$(( SECTOR_SIZE * SECTOR_TOTAL ))" || return 1
+
+    echo "Writing partition table:"
+    cat "${partition_table_file}"
+    chroot "${rootfs_dir}" /sbin/sfdisk "${STORAGE_DEVICE_FILE}" < "${partition_table_file}" || return 1
+
+    format_dummy_partition "/sbin/mkfs.ext4" "$(( SECTOR_SIZE * BOOT_START ))" "${STORAGE_DEVICE_FILE}" || return 1
+    format_dummy_partition "/usr/sbin/mkfs.f2fs" "$(( SECTOR_SIZE * ROOTFS_START ))" "${STORAGE_DEVICE_FILE}" || return 1
+    format_dummy_partition "/usr/sbin/mkfs.f2fs" "$(( SECTOR_SIZE * SPOOL_START ))" "${STORAGE_DEVICE_FILE}" || return 1
+
+    unlink "${partition_table_file}"
```

Comment:
because you create this in this function, i understand why you clean it up at the end again. However if any of the previous commands fail, this will never get cleaned.